### PR TITLE
feat: add withIndex operations

### LINKED ISSUE #7: filterWithIndex or xxWithIndex
**Is your feature request related to a problem? Please describe.**
When using `filter`, the block only contains item but not index. 
For example, I using an array as a EZS_Sequence, it has 10 items. Now I want to get all items which index is lower then 5. `filter` can't do this.

**Describe the solution you'd like**
Not only `filter`, If you can get the index, you should better provide a version that include 'WithIndex:'

### PR DESCRIPTION
**Pull request by face2d**
fixes #7 
- filterWithIndex
- selectWithIndex
- rejectWithIndex

### CODE REVIEW

**Comment by ValiantCat** [comment]
commit message  location

**Comment by WilliamZang** [comment]
Please increase test coverage.

File: EasySequence/Classes/Operations/EZSequence+Operations.m [hunk 1]
```diff
@@ -94,6 +94,18 @@ - (EZSequence *)filter:(EZSFliterBlock)filterBlock {
     return EZS_Sequence(array);
 }
 
+- (EZSequence *)filterWithIndex:(EZSFilterWithIndexBlock)filterBlock {
+    NSParameterAssert(filterBlock);
+    NSMutableArray *array = [NSMutableArray new];
+    NSUInteger index = 0;
+    for (id item in self) {
+        if (filterBlock && filterBlock(item, index++)) {
```
[the remark this answers is not in the document]
Comment (face2d) [reply EasySequence/Classes/Operations/EZSequence+Operations.m]: fixed

File: EasySequence/Classes/Operations/EZSequence+Operations.h [hunk 2]
```diff
@@ -76,6 +76,14 @@ FOUNDATION_EXTERN NSString * const EZSequenceExceptionReason_ZipMethodMustUseOnN
  */
 - (EZSequence<T> *)filter:(BOOL (NS_NOESCAPE ^)(T item))filterBlock;
 
+/**
+ Returns a new sequence containing all elements of self for which the given block returns a true value.
+ 
+ @param filterBlock A block to filter elements
```
[the remark this answers is not in the document]
Comment (face2d) [reply EasySequence/Classes/Operations/EZSequence+Operations.h]: fixed

File: EasySequence/Classes/Operations/EZSequence+Operations.h [hunk 3]
```diff
@@ -162,6 +178,14 @@ FOUNDATION_EXTERN NSString * const EZSequenceExceptionReason_ZipMethodMustUseOnN
  */
 - (EZSequence<T> *)reject:(BOOL (NS_NOESCAPE ^)(T item))rejectBlock;
 
+/**
+ Returns a new sequence containing the elements in self for which the given block is not true
+ 
+ @param rejectBlock A block to reject element
```
[the remark this answers is not in the document]
Comment (face2d) [reply EasySequence/Classes/Operations/EZSequence+Operations.h]: fixed

[the remark this answers is not in the document]
Comment (face2d) [reply EasySequence/Classes/Operations/EZSequence+Operations.h]: fixed

**Pull request merged**